cosmos/providers/dbt/core/profiles/databricks.py

```diff
@@ -1,13 +1,45 @@
+from typing import TYPE_CHECKING, Optional, Tuple
+
+if TYPE_CHECKING:
+    from airflow.models import Connection
+
 databricks_profile = {
     "outputs": {
         "dev": {
             "type": "databricks",
             "host": "{{ env_var('DATABRICKS_HOST') }}",
             "catalog": "{{ env_var('DATABRICKS_CATALOG') }}",
-            "schema": "{{ env_var('DATABRICKS_DATABASE') }}",
+            "schema": "{{ env_var('DATABRICKS_SCHEMA') }}",
```

Comment:
@pohek321 do you remember if there was a reason we called this `DATABRICKS_SCHEMA` instead of `DATABRICKS_DATABASE`?

Reply:
yes, in databricks-land a database is really a schema and a catalog is a database. Super confusing - I'm putting together a commit to fix and will push it up momentarily.

Reply:
Additionally, I was mirroring all other conns in Airflow that use `schema` as a db parameter (with the exception of Snowflake) -- so that's why it's this way. The commit I'm putting together will also reflect that.

Reply:
The complication that I was having was that the "database" is actually a Unity Catalog which is optional and there for the purposes of data lineage and asset management [docs](https://www.databricks.com/product/unity-catalog).
With dbt it looks like it was introduced as a seperate config item in [v1.4.0](https://github.com/databricks/dbt-databricks/blob/main/CHANGELOG.md#dbt-databricks-140-january-25-2023). Prior to that it was acceptable to use dot seperation in the schema field.

We're not currently using the Unity Catalog and our schemas are AWS Glue databases which reference s3 locations.

As it's optional, if it's not populated then there will be a dependency on fix #134

Reply:
@jlaneve & @MikeWallis42 can you both take a look at my [last commit](https://github.com/astronomer/astronomer-cosmos/pull/140/commits/902fd45fe27de878807dc9d3a6be48f1e8dbed11)? Our connectivity DAGs were breaking before making this change and this resolved it. Here's what I did purely based on Airflow connection patterns and what's standard in Airflow:

- BigQuery: 
  - Project (DB): use the `database_override` first, if none is specified, then use the `project_id` from the json creds
  - Dataset (Schema): require a `dataset` to be specified via the `schema` parameter. If the schema parameter isn't passed - fail the task
- Databricks 
  - Catalog (DB): use the `database_override` first, and the `conn.schema` second. If neither of those is specified default to the value `hive_metastore` (more details [here](https://docs.databricks.com/data-governance/unity-catalog/hive-metastore.html#default-catalog))
  - Database (Schema): use the `schema_override` for the Databricks Database name to follow Airflow connection parameter patterns. If the schema isn't passed - fail the task
- Postgres: 
  - Database (DB): use the `database_override` first, and the `conn.schema` second. If neither of those is specified then fail
  - Schema (Schema): use the `schema_override`, if the schema parameter isn't passed - fail the task. 
- Redshift:
  - Database (DB): use the `database_override` first, and the `conn.schema` second. If neither of those is specified then fail
  - Schema (Schema): use the `schema_override`, if the schema parameter isn't passed - fail the task.

Open to having a discussion to change them, but since these workflows are being executed *in Airflow* I think it makes the most sense to follow Airflow patterns - i.e. Connection Schema parameter is traditionally used for a database.

Reply:
@pohek321 I think my issue with the databricks one is that the Unity catalog is optional and isn't enabled by default, [ref](https://docs.databricks.com/data-governance/unity-catalog/get-started.html#overview-of-unity-catalog-setup).
Also as per this [page](https://docs.databricks.com/data-governance/unity-catalog/index.html#the-unity-catalog-object-model)

> Catalog: The first layer of the object hierarchy, used to organize your data assets.
> Schema: Also known as databases, schemas are the second layer of the object hierarchy and contain tables and views.

So the schema is a database and catalog is a higher level of grouping.

Would it be better to define it as:-

1. Schema will be the level above an object such as table/view.
 - So in databricks the Airflow connection Schema will be the Databricks Schema
 - For BigQuery the Airflow connection Schema would be a Dataset. 
 - RDBMS: Schema = Schema.
2. Database will be the level above Schema.
 - Databricks: Unity Catalog.
 - BigQuery: Project
 - RDBMS: Database = Database

Is it the case that these connections to an RDBMS would normally use schema for database and then you'd pass the schema in the table name like `my_schema.my_table`?

Reply:
@MikeWallis42 

> So in databricks the Airflow connection Schema will be the Databricks Schema

I think you meant: So in databricks the Airflow connection Schema will be the Databricks *Database*

This is counterintuitive to all other Airflow connection objects because there isn't a single Airflow connection object that uses the `Schema` parameter for an actual schema (except Snowflake). So, this would be an anti-pattern if we defined it differently for Databricks. Databricks docs state that if a catalog *isn't* specified, then `hive_metastore` is used by default (see [here](https://docs.databricks.com/data-governance/unity-catalog/hive-metastore.html#default-catalog)). Our users could overwrite that when instantiating the cosmos class like this:

```python
  jaffle_shop = DbtTaskGroup(
      ...,
      conn_id="databricks_default",
      dbt_args={
          "db_name": "databricks_catalog" # overrides the `hive_metastore` default
          "schema": "databricks_db" # sets a default schema (or databricks database) - the value specified here can be overridden in dbt .yml/.sql files
      },
  )
```

Additionally, dbt treats `catalog.database` databricks objects the same way it treats `db.schema` rds objects. So if you specify the following in a dbt properties.yml:

```yaml
version: 2

models:
  - name: customers
    description: This table has basic information about a customer, as well as some derived facts based on a customer's orders
    config:
      schema: test_schema
```

It will create a be and new "Databricks Database" called `test_schema` like so:
![Screenshot 2023-02-15 at 10 04 27 AM](https://user-images.githubusercontent.com/31361051/219101289-8cfac058-04cc-4d2a-86a0-12ee0bb6d591.png)

So, in summary - we should keep the pattern that Airflow is following, doing so will also align with how dbt treats these objects:

(RDS `db.schema.table` = Databricks `catalog.database.table` = BigQuery `project_id.dataset.table`

Reply:
@pohek321 thanks for getting back to me so early your time!
ok so basically my original issue should have just been a documentation update. It's quite confusing when you see database and schema being used interchangably across implementations.
- Databricks UI: database
- dbt: schema
- Airflow:
  - Connections; N/A but typically schema is a database.
  - Operators; expects you to pass them in as arguments but uses the terminology catalog and schema.
- cosmos: Pass in `{"dbt_args": {"schema": "my_schema"}}`

So basically there are two methods of using dbt and cosmos to write to a specific databricks database. Schema in dbt_args or like [this](https://docs.getdbt.com/reference/resource-configs/schema) as you showed above.
It will not be possible to add something on the Airflow connection to default this as the databricks connection doesn't to this date.
I doubt I'll be the first to question this so I'm happy to go with the desired implementation just as long as it's documented  😄

Reply:
@MikeWallis42 

Awesome, I went ahead and added a commit to bring clarity to all of this in our docs. See commit [0f37905](https://github.com/astronomer/astronomer-cosmos/pull/140/commits/0f37905cc4f0a0476c4389303622417435f0382c)

Reply:
Marking as resolved and approving.